File: android/DoerApp/app/src/main/java/com/doer/app/ui/tasks/TaskViewModel.kt
package com.doer.app.ui.tasks

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.doer.app.data.model.DoerTask
import com.doer.app.data.model.TaskWithLabels
import com.doer.app.data.repository.TaskRepository
import com.doer.app.util.DateHelpers
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class TaskUiState(
    val tasks: List<TaskWithLabels> = emptyList(),
    val todayTasks: List<TaskWithLabels> = emptyList(),
    val overdueTasks: List<TaskWithLabels> = emptyList(),
    val upcomingByDay: Map<String, List<TaskWithLabels>> = emptyMap(),
    val isLoading: Boolean = false,
    val error: String? = null
)

class TaskViewModel : ViewModel() {

    private val taskRepository = TaskRepository()

    private val _uiState = MutableStateFlow(TaskUiState())
    val uiState: StateFlow<TaskUiState> = _uiState.asStateFlow()

    fun fetchByProject(projectId: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val tasks = taskRepository.fetchByProject(projectId)
                _uiState.update { it.copy(tasks = tasks, isLoading = false) }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun fetchForToday() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val all = taskRepository.fetchForToday()
                val today = DateHelpers.todayString()
                val overdue = all.filter { DateHelpers.isOverdue(it.dueDate) }
                val todayOnly = all.filter { it.dueDate?.take(10) == today }
                _uiState.update {
                    it.copy(
                        todayTasks = todayOnly,
                        overdueTasks = overdue,
                        isLoading = false
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun fetchForUpcoming() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val tasks = taskRepository.fetchForUpcoming()
                val grouped = tasks.groupBy { it.dueDate?.take(10) ?: "No date" }
                    .toSortedMap()
                _uiState.update {
                    it.copy(upcomingByDay = grouped, isLoading = false)
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(isLoading = false, error = e.message) }
            }
        }
    }

    fun createTask(
        userId: String,
        title: String,
        projectId: String,
        sectionId: String? = null,
        priority: Int = 4,
        dueDate: String? = null
    ) {
        viewModelScope.launch {
            try {
                val currentTasks = _uiState.value.tasks
                val maxPosition = currentTasks.maxOfOrNull { it.position } ?: 0.0
                taskRepository.create(
                    userId = userId,
                    title = title,
                    projectId = projectId,
                    sectionId = sectionId,
                    priority = priority,
                    dueDate = dueDate,
                    position = maxPosition + 65536.0
                )
                // Refresh whichever view is active
                fetchByProject(projectId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun toggleComplete(task: DoerTask) {
        viewModelScope.launch {
            try {
                taskRepository.toggleComplete(task)
                // Remove from current list optimistically
                _uiState.update { state ->
                    state.copy(
                        tasks = state.tasks.filter { it.id != task.id },
                        todayTasks = state.todayTasks.filter { it.id != task.id },
                        overdueTasks = state.overdueTasks.filter { it.id != task.id },
                        upcomingByDay = state.upcomingByDay.mapValues { (_, v) ->
                            v.filter { it.id != task.id }
                        }.filterValues { it.isNotEmpty() }
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun deleteTask(taskId: String) {
        viewModelScope.launch {
            try {
                taskRepository.delete(taskId)
                _uiState.update { state ->
                    state.copy(
                        tasks = state.tasks.filter { it.id != taskId },
                        todayTasks = state.todayTasks.filter { it.id != taskId },
                        overdueTasks = state.overdueTasks.filter { it.id != taskId },
                        upcomingByDay = state.upcomingByDay.mapValues { (_, v) ->
                            v.filter { it.id != taskId }
                        }.filterValues { it.isNotEmpty() }
                    )
                }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun updateTask(taskId: String, fields: Map<String, Any?>) {
        viewModelScope.launch {
            try {
                taskRepository.update(taskId, fields)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun addLabel(taskId: String, labelId: String) {
        viewModelScope.launch {
            try {
                taskRepository.addLabel(taskId, labelId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun removeLabel(taskId: String, labelId: String) {
        viewModelScope.launch {
            try {
                taskRepository.removeLabel(taskId, labelId)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.message) }
            }
        }
    }

    fun clearError() {
        _uiState.update { it.copy(error = null) }
    }
}

File: ios/DoerApp/DoerApp/ViewModels/TaskViewModel.swift
import SwiftUI
import Supabase

@MainActor
final class TaskViewModel: ObservableObject {
    @Published var tasks: [DoerTask] = []
    @Published var isLoading = false
    @Published var errorMessage: String?

    private let taskService = TaskService()
    private var loadedContext: String?

    // MARK: - Fetch

    func fetchByProject(projectId: String, force: Bool = false) async {
        guard force || loadedContext != "project:\(projectId)" else { return }
        if tasks.isEmpty { isLoading = true }
        defer { isLoading = false }
        do {
            tasks = try await taskService.fetchByProject(projectId: projectId)
            loadedContext = "project:\(projectId)"
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func fetchForToday(force: Bool = false) async {
        guard force || loadedContext != "today" else { return }
        if tasks.isEmpty { isLoading = true }
        defer { isLoading = false }
        do {
            tasks = try await taskService.fetchForToday()
            loadedContext = "today"
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func fetchForUpcoming(force: Bool = false) async {
        guard force || loadedContext != "upcoming" else { return }
        if tasks.isEmpty { isLoading = true }
        defer { isLoading = false }
        do {
            tasks = try await taskService.fetchForUpcoming()
            loadedContext = "upcoming"
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func fetchSubTasks(parentId: String) async -> [DoerTask] {
        do {
            return try await taskService.fetchSubTasks(parentId: parentId)
        } catch {
            errorMessage = error.localizedDescription
            return []
        }
    }

    func invalidate() {
        loadedContext = nil
    }

    // MARK: - Create

    func createTask(
        title: String,
        projectId: String,
        sectionId: String? = nil,
        parentTaskId: String? = nil,
        priority: Int = 4,
        dueDate: String? = nil
    ) async {
        guard let userId = await AuthService().getUser()?.id.uuidString else { return }

        let lastPosition = tasks
            .filter { $0.sectionId == sectionId && $0.parentTaskId == parentTaskId }
            .map(\.position)
            .max()
        let position = PositionHelpers.getPositionAtEnd(lastPosition: lastPosition)

        do {
            let newTask = try await taskService.create(
                userId: userId,
                title: title,
                projectId: projectId,
                sectionId: sectionId,
                parentTaskId: parentTaskId,
                priority: priority,
                dueDate: dueDate,
                position: position
            )
            tasks.append(newTask)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Toggle / Delete

    func toggleComplete(task: DoerTask) async {
        if let index = tasks.firstIndex(where: { $0.id == task.id }) {
            tasks[index].isCompleted.toggle()
            if tasks[index].isCompleted {
                let taskId = task.id
                try? await Task.sleep(nanoseconds: 500_000_000)
                tasks.removeAll { $0.id == taskId }
            }
        }
        do {
            try await taskService.toggleComplete(task: task)
        } catch {
            if let index = tasks.firstIndex(where: { $0.id == task.id }) {
                tasks[index].isCompleted.toggle()
            }
            errorMessage = error.localizedDescription
        }
    }

    func deleteTask(id: String) async {
        let removed = tasks.first { $0.id == id }
        tasks.removeAll { $0.id == id }
        do {
            try await taskService.delete(id: id)
        } catch {
            if let removed {
                tasks.append(removed)
            }
            errorMessage = error.localizedDescription
        }
    }

    func updateTask(id: String, fields: TaskUpdate) async {
        do {
            try await taskService.update(id: id, fields: fields)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func addLabel(taskId: String, labelId: String) async {
        do {
            try await taskService.addLabel(taskId: taskId, labelId: labelId)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    func removeLabel(taskId: String, labelId: String) async {
        do {
            try await taskService.removeLabel(taskId: taskId, labelId: labelId)
        } catch {
            errorMessage = error.localizedDescription
        }
    }

    // MARK: - Computed

    func tasksForSection(projectId: String, sectionId: String?) -> [DoerTask] {
        tasks.filter { $0.projectId == projectId && $0.sectionId == sectionId && $0.parentTaskId == nil }
            .sorted { $0.position < $1.position }
    }

    var todayTasks: [DoerTask] {
        let today = DateHelpers.todayString()
        return tasks.filter { $0.dueDate == today && !$0.isCompleted }
    }

    var overdueTasks: [DoerTask] {
        tasks.filter { DateHelpers.isOverdue($0.dueDate) && !$0.isCompleted }
    }

    var upcomingTasksByDay: [(date: String, tasks: [DoerTask])] {
        let grouped = Dictionary(grouping: tasks.filter { !$0.isCompleted }) { $0.dueDate ?? "" }
        return grouped
            .sorted { $0.key < $1.key }
            .map { (date: $0.key, tasks: $0.value) }
    }
}
